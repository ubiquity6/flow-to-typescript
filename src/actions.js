// @flow

import { makeAction } from './ActionEmitter';
import type { Action } from './ActionEmitter';
import type { UpdateStep } from './Inputs';


export const LifecycleActions = {
  INITIALIZE: (makeAction('LIFECYCLE_ACTION_INITIALIZE'): Action<void>),
  PRE_UPDATE: (makeAction('LIFECYCLE_ACTION_PRE_UPDATE'): Action<UpdateStep>),
  FIXED_UPDATE: (makeAction('LIFECYCLE_FIXED_UPDATE'): Action<UpdateStep>),
  UPDATE: (makeAction('LIFECYCLE_ACTION_UPDATE'): Action<UpdateStep>),
  POST_UPDATE: (makeAction('LIFECYCLE_ACTION_POST_UPDATE'): Action<UpdateStep>),
};

const test = [2,3,
4,4,5];