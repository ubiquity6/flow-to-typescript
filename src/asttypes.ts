
export type ASTEntry = {
    type: string,
    range: any,
    loc: any
}

export type TypeCastExpression = ASTEntry & {
    expression: ASTEntry,
    typeAnnotation: ASTEntry
}


export type TypeParameter = ASTEntry & {
    type: "TypeParameter",
    bound: ASTEntry,
    variance: ASTEntry,
    default: ASTEntry
}

export type TSTypeParameter = ASTEntry & {
    type: "TSTypeParameter",
    constraint: ASTEntry,
}

export type Identifier = ASTEntry & {
    type: "Identifier",
    name: string,
    typeAnnotation: ASTEntry,
    optional: boolean
}

export type ObjectTypeProperty = ASTEntry & {
    type: "ObjectTypeProperty",
    key: ASTEntry,
    value: ASTEntry,
    method: boolean,
    optional: boolean,
    static: boolean,
    proto: boolean
    variance: ASTEntry
    kind: any
}

export type ImportSpecifier = ASTEntry & {
    type: "ImportSpecifier",
    imported: ASTEntry,
    local: ASTEntry,
    importKind: "type" | "value"
}



export type FunctionTypeAnnotation = ASTEntry & {
    type: "FunctionTypeAnnotation",
    params: ASTEntry[],
    returnType: ASTEntry,
    rest: any, // ?
    typeParameters: any// ?
}

export type InterfaceDeclaration = ASTEntry & {
    type: "InterfaceDeclaration",
    id: ASTEntry,
    typeParameters: Array<ASTEntry>, //?
    body: ObjectTypeAnnotation
    extends: Array<ASTEntry>
}

export type ObjectTypeAnnotation = ASTEntry & {
    type: "ObjectTypeAnnotation",
    exact: boolean,
    properties: Array<ASTEntry>
    indexers: Array<ASTEntry>,
    callProperties: Array<ASTEntry>,
    internalSlots: Array<ASTEntry>
}
