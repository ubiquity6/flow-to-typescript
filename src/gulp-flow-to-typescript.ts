import { ConvertOptions, convert } from ".";
var PluginError = require('plugin-error');

var PLUGIN_NAME = 'gulp-flow-to-typescipt';

import {Transform} from 'readable-stream';
import path from 'path';

export function gulp(options: ConvertOptions) {
    if (!options) {
        options = {};
    }

    return new Transform({
        objectMode: true,
        transform: function (file, enc, callback) {
            if (file.isNull()) {
                return callback(null, file);
            }

            function doReplace(this: any) {
                if (file.isStream()) {
                    this.emit('error', new PluginError(PLUGIN_NAME, 'Streams not supported!'));

                }

                if (file.isBuffer()) {
                    const parsedpath = path.parse(file.path);
                    console.log("PARSING:", file.path, parsedpath.ext);
                      
                    if(parsedpath.ext === '.js') {
                        // assume it is a flow file.
                     //   console.log("PARSING:", file.path);
                        const outtxt= convert(String(file.contents), file.filepath, {
                            no_emit: true
                        });
                        file.contents = new Buffer(outtxt);


                        //let filePath = path.parse(file.path);
                        //filePath.base = filePath.base.replace(path.extname(file.path), 'ts');
                        // format the path back into an absolue
                        //file.path = path.format(filePath);
                        
                        parsedpath.ext = '.ts';
                        delete parsedpath.base;
                        //console.log(parsedpath);
                        file.path = path.format(parsedpath);
                        //console.log("RESULT:", file.path);
                    }
                   
                    callback(null, file);
                }
            }

            doReplace();
        }
    });
};
