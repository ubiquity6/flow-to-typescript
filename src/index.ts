import * as prettier from 'prettier';
import fs from 'fs';

export type convertOptions = {
  dump_ast? : boolean
  no_emit? : boolean
  print_types?: boolean
}
export function convert(srcpath: string, options : convertOptions) {


  let srcast: any;
  var parserSrc: prettier.CustomParser = (text : string, { flow }) => {
    const ast : prettier.AST = flow(text);

    // convert "type" imports to normal imports. todo: improve: use /// <reference> which is the actual equivalent
    ast.body.filter(e => e.type === "ImportDeclaration" && e.importKind === "type").forEach(e => e.importKind = "value");

    recurseFindTypes(ast.body, {TypeCastExpression: removeTypeCastExpressions});
    srcast = ast;

    return ast;
  }

  type ASTEntry = {
    type: string,
    range: any,
    loc: any
  }

  type TypeCastExpression = ASTEntry &  {
    expression : ASTEntry,
    typeAnnotation: ASTEntry
  }

  function recurseFindTypes(obj: ASTEntry, visitors: {[key:string] : (obj:ASTEntry) => void}) {
    if(options.print_types)
      console.log("type:", obj.type)
    for(var k in obj) {
      const child = obj[k];
      if(child && child.type) {
        if(visitors[child.type]) {
          visitors[child.type](child);
        }
        recurseFindTypes(child as ASTEntry, visitors);      
      }
      if(child && typeof child === 'object' && child[0] !== undefined) { // is array 
        child.filter(o=> o.type).forEach(o => recurseFindTypes(o, visitors));
      }
    }
  }

  function removeTypeCastExpressions(o: ASTEntry) {
    const obj = o as TypeCastExpression;
    // substitute body with contents of expression.
    delete obj.typeAnnotation;
    const expression = obj.expression;
    delete obj.expression;
    Object.assign(obj, expression);
  }

  const srctxt = fs.readFileSync(srcpath, 'utf-8');

  // capture flow ast
  const result = prettier.format(srctxt, {
    parser : parserSrc
  });

  if(options.dump_ast) {
    fs.writeFileSync(srcpath.replace(".js", ".ast"), JSON.stringify(srcast, null, 4));
  }

  if(!options.no_emit) {
    fs.writeFileSync(srcpath.replace(".js", ".ts"), result);
  }

  return result;
}

