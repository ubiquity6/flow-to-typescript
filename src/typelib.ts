import * as ts from "typescript";

export function writeDeclaration(source: string, outfile: string) {
    let result = ts.transpileModule(source, {
        compilerOptions: { 
            module: ts.ModuleKind.CommonJS,
            declaration: true,
            emitDeclarationOnly: true,
            outFile: outfile
           }
      });
      console.log(JSON.stringify(result));
      
}