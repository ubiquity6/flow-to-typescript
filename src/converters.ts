import { ASTEntry, TypeCastExpression, TypeParameter, TSTypeParameter, ObjectTypeProperty, ImportSpecifier, Identifier, FunctionTypeAnnotation } from "./asttypes";

export function removeTypeCastExpressions(o: ASTEntry) {
    const obj = o as TypeCastExpression;
    // substitute body with contents of expression.
    delete obj.typeAnnotation;
    const expression = obj.expression;
    delete obj.expression;
    Object.assign(obj, expression);
}

export function removeTypeParameterDefault(e: ASTEntry) {
    const obj = e as TypeParameter;
    delete obj.default;
}

export function convertTypeTypeAnnotationToTSTypeReference(e: ASTEntry): ASTEntry {
    if (e.type !== 'TypeAnnotation') {
        throw new Error("wrong");
    }
    const r = (e as any).typeAnnotation;
    if (r.type !== 'GenericTypeAnnotation') {
        console.log("convertTypeTypeAnnotationToTSTypeReference: dont know this one:", r.type);
    }
    r.type = "TSTypeReference";
    r.typeName = r.id;
    delete r.id;
    return r;
}

export function convertTypeParameter(e: ASTEntry) {
    const obj = e as TypeParameter;
    const ts = e as TSTypeParameter;
    removeTypeParameterDefault(e); // don't think typescript supports default args for generics.
    e.type = "TSTypeParameter";
    if (obj.bound) {
        ts.constraint = obj.bound;
        ts.constraint = convertTypeTypeAnnotationToTSTypeReference(obj.bound);
        delete obj.bound;
    }
}

export function convertTypeAnnotation(e: ASTEntry) {
    e.type = "TSTypeAnnotation";
}

export function convertTypeCastExpression(e: ASTEntry) {
    e.type = "TSAsExpression";
}

export function convertAnyTypeAnnotation(e: ASTEntry) {
    e.type = "TSAnyKeyword";
}

export function convertVoidTypeAnnotation(e: ASTEntry) {
    e.type = "TSVoidKeyword";
}

export function convertObjectTypeProperty(e: ASTEntry) {
    var ot = e as ObjectTypeProperty
    // right now, we just want to strip variance.
    delete ot.variance;

    // check if it's a constructor signature
    if(ot.key && ot.key.type == "Identifier") {
        var id = ot.key as Identifier;
        if(id.name === "constructor" && ot.value.type === "FunctionTypeAnnotation") {
            // flow likes to put void return types on constructors.
            // typescript does not.
            var ctorFunction = ot.value as FunctionTypeAnnotation;
            if(ctorFunction.returnType.type === "VoidTypeAnnotation")
                delete ctorFunction.returnType;
        }
    }
}

export function convertImportSpecifier(e: ASTEntry) {
    var ot = e as ImportSpecifier
    // typescript doens't do type only imports in this way.
    if (ot.importKind === "type") {
        delete ot.importKind;
    }
}
