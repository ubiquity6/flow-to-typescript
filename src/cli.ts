import commandLineArgs from 'command-line-args';
import { convert } from './index';
import * as fs from 'fs';
import * as path from 'path';
import { writeDeclaration } from './typelib';

const optionDefinitions = [
//  { name: 'verbose', alias: 'v', type: Boolean },
  { name: 'src', type: String, multiple: true, defaultOption: true },
  { name: 'emitSrc', type: Boolean, defaultValue: true },
  { name: 'emitTypes', type: Boolean, defaultValue: true }
//  { name: 'timeout', alias: 't', type: Number }
]

const options = commandLineArgs(optionDefinitions)

console.log("options:", options);

if(options.src) {
    options.src.forEach((src:string) => {
        console.log("processing ", src);
        var tsSource = convert(src, {
            no_emit : !options.emitSrc
        });

        if(options.emitTypes) {
            console.log("emitting types");
            writeDeclaration(tsSource, src.replace('.js', 'd.ts'));
        }
    })
}
else {
       
}
