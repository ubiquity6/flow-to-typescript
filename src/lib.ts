import * as prettier from 'prettier';
import fs from 'fs';

import {ASTEntry}  from './asttypes';
import { convertTypeCastExpression, convertTypeParameter, convertTypeAnnotation, convertAnyTypeAnnotation, convertVoidTypeAnnotation, convertObjectTypeProperty, convertImportSpecifier } from './converters';

export type ConvertOptions = {
  dump_ast? : boolean
  no_emit? : boolean
  print_types?: boolean
}
/**
 * 
 * @param srctxt {string} source code. pass in null to just load from srcpath
 * @param srcpath {string} source path. used to figure out name of file.
 * @param options {ConvertOptions}
 */
export function convert(srctxt:string|null, srcpath: string, options : ConvertOptions) {


  var parserSrc: prettier.CustomParser = (text : string, { flow }) => {
    const ast : prettier.AST = flow(text);

    // convert "type" imports to normal imports. todo: improve: use /// <reference> which is the actual equivalent
    ast.body.filter(e => e.type === "ImportDeclaration" && e.importKind === "type").forEach(e => e.importKind = "value");
    ast.body.filter(e => e.type === "ExportNamedDeclaration" && e.exportKind === "type").forEach(e => e.exportKind = "value");
    
    recurseFindTypes(ast.body, {
      TypeCastExpression: convertTypeCastExpression,
//      TypeCastExpression: removeTypeCastExpressions,
      TypeParameter: convertTypeParameter,
      TypeAnnotation: convertTypeAnnotation,
      AnyTypeAnnotation: convertAnyTypeAnnotation,
      VoidTypeAnnotation: convertVoidTypeAnnotation,
      ObjectTypeProperty: convertObjectTypeProperty,
      ImportSpecifier: convertImportSpecifier
    });

    

    if(options.dump_ast) {
      fs.writeFileSync(srcpath.replace(".js", ".ast"), JSON.stringify(ast, null, 4));
    }

    return ast;
  }

  function recurseFindTypes(obj: ASTEntry, visitors: {[key:string] : (obj:ASTEntry) => void}) {
    if(options.print_types)
      console.log("type:", obj.type, " at ", obj.loc? obj.loc.start.line : "unknown")
    if(visitors[obj.type]) {
      visitors[obj.type](obj);
    }
    for(var k in obj) {
      const child = obj[k];
      if(child && child.type) {
        recurseFindTypes(child as ASTEntry, visitors);      
      }
      if(child && typeof child === 'object' && child[0] !== undefined) { // is array 
        child.filter(o=> o.type).forEach(o => recurseFindTypes(o, visitors));
      }
    }
  }


  // legacy api
  if(!srctxt) {
    srctxt = fs.readFileSync(srcpath, 'uft-8');
  }


  // capture flow ast
  const result = prettier.format(srctxt, {
    parser : parserSrc
  });

  if(!options.no_emit) {
    fs.writeFileSync(srcpath.replace(".js", ".ts"), result);
  }

  return result;
}


export function analyseref(srcpath: string, options : ConvertOptions) {


  var parserSrc: prettier.CustomParser = (text : string, { typescript }) => {
    const ast : prettier.AST = typescript(text);

    if(options.dump_ast) {
      fs.writeFileSync(srcpath.replace(".ts", ".ast"), JSON.stringify(ast, null, 4));
    }

    return ast;
  }

  const srctxt = fs.readFileSync(srcpath, 'utf-8');

  // capture typescript ast
  const result = prettier.format(srctxt, {
    parser : parserSrc
  });

  return result;
}
