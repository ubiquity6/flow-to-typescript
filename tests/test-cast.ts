
import { convert,analyseref } from '../src/index';

describe('cast', () => {
    it('main', () =>
      expect(convert(null, 'tests/src/cast.js', {dump_ast: true, 
        no_emit: true,
        print_types: false
    })).toMatchSnapshot()
  )
});

// generate ast for comparison and debugging
var result = analyseref('tests/out/cast.ts', {dump_ast: true, 
    no_emit: true,
    print_types: false
});