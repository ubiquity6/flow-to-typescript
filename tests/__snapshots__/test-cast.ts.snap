// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cast main 1`] = `
"function addChild(child: TransformComponent): void {
  const tc = (mutableState.parent as any) as TransformComponent;
}
"
`;
